spring4/src/test/java/org/jdbi/v3/spring4/TestJdbiTemplate.java

```diff
@@ -0,0 +1,151 @@
+                if (count == 1) {
+                    throw new ForceRollback();
+                } else {
+                    throw new RuntimeException("!ZABAK");
+                }
+            });
+        });
+
+        final int count = jdbiOps.withHandle(h -> h.createQuery("select count(*) from something").mapTo(Integer.class).findOnly());
+        assertThat(count).isEqualTo(0);
+    }
+
+    @Test
+    public void testNested() throws Exception {
+        assertThatExceptionOfType(ForceRollback.class).isThrownBy(() -> {
+            service.inPropagationRequired(outer -> {
+                outer.execute("insert into something (id, name) values (7, 'ignored')");
+
+                assertThatExceptionOfType(ForceRollback.class).isThrownBy(() -> {
+                    service.inNested(inner -> {
+                        inner.execute("insert into something (id, name) values (8, 'ignored again')");
+
+                        int count = inner.createQuery("select count(*) from something").mapTo(Integer.class).findOnly();
+                        assertThat(count).isEqualTo(2);
+                        throw new ForceRollback();
+                    });
+                });
+                int count = outer.createQuery("select count(*) from something").mapTo(Integer.class).findOnly();
+                assertThat(count).isEqualTo(1);
+                throw new ForceRollback();
+            });
+        });
+        service.inPropagationRequired(h -> {
+            int count = h.createQuery("select count(*) from something").mapTo(Integer.class).findOnly();
+            assertThat(count).isEqualTo(0);
+        });
+    }
+
+    @Test
+    public void testRequiresNew() throws Exception {
+        service.inPropagationRequired(outer -> {
+            outer.execute("insert into something (id, name) values (7, 'ignored')");
+
+            assertThatExceptionOfType(ForceRollback.class).isThrownBy(() -> {
+                service.inRequiresNewReadUncommitted(inner -> {
+                    int count = inner.createQuery("select count(*) from something").mapTo(Integer.class).findOnly();
+                    assertThat(count).isEqualTo(1);
+                    inner.execute("insert into something (id, name) values (8, 'ignored again')");
+                    throw new ForceRollback();
+                });
+            });
+
+            int count = outer.createQuery("select count(*) from something").mapTo(Integer.class).findOnly();
+            assertThat(count).isEqualTo(1);
+        });
+    }
+
+    @Configuration
+    @EnableTransactionManagement
+    public static class Config {
+
+        @Bean
+        public DataSource dataSource() {
+            return new EmbeddedDatabaseBuilder()
+                    .setType(EmbeddedDatabaseType.H2)
+                    .build();
+        }
+
+        @Bean
+        public Jdbi jdbi(DataSource dataSource) {
+            return Jdbi.create(dataSource);
```

Comment:
What if this instead created a `DataSouce` implementation that does the `JdbiTemplate` / `JdbiUtil` logic, and delegates to the `dataSource` parameter?

Reply:
That's okay, but see my comment above. I think in this case spring manipulates stuff and state which Jdbi does not know anything about. also other problems I mentioned arise.